provider/webhook/webhook.go

```diff
@@ -0,0 +1,263 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/plan"
+
+	backoff "github.com/cenkalti/backoff/v4"
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	mediaTypeFormatAndVersion = "application/external.dns.webhook+json;version=1"
+	contentTypeHeader         = "Content-Type"
+	acceptHeader              = "Accept"
+	maxRetries                = 5
+)
+
+var (
+	recordsErrorsGauge = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: "external_dns",
+			Subsystem: "webhook_provider",
+			Name:      "records_errors",
+			Help:      "Errors with Records method",
+		},
+	)
+	applyChangesErrorsGauge = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: "external_dns",
+			Subsystem: "webhook_provider",
+			Name:      "applychanges_errors",
+			Help:      "Errors with ApplyChanges method",
+		},
+	)
+	adjustEndpointsErrorsGauge = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: "external_dns",
+			Subsystem: "webhook_provider",
+			Name:      "adjustendpointsgauge_errors",
+			Help:      "Errors with AdjustEndpoints method",
+		},
+	)
+)
+
+type WebhookProvider struct {
+	client          *http.Client
+	remoteServerURL *url.URL
+	DomainFilter    endpoint.DomainFilter
+}
+
+func init() {
+	prometheus.MustRegister(recordsErrorsGauge)
+	prometheus.MustRegister(applyChangesErrorsGauge)
+	prometheus.MustRegister(adjustEndpointsErrorsGauge)
+}
+
+func NewWebhookProvider(u string) (*WebhookProvider, error) {
+	parsedURL, err := url.Parse(u)
+	if err != nil {
+		return nil, err
+	}
+
+	// negotiate API information
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set(acceptHeader, mediaTypeFormatAndVersion)
+
+	client := &http.Client{}
+	var resp *http.Response
+	err = backoff.Retry(func() error {
+		resp, err = client.Do(req)
+		if err != nil {
+			log.Debugf("Failed to connect to plugin api: %v", err)
+			return err
+		}
+		// we currently only use 200 as success, but considering okay all 2XX for future usage
+		if resp.StatusCode >= 300 && resp.StatusCode < 500 {
+			return backoff.Permanent(fmt.Errorf("status code < 500"))
+		}
+		return nil
+	}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxRetries))
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to plugin api: %v", err)
+	}
+
+	contentType := resp.Header.Get(contentTypeHeader)
+
+	// read the serialized DomainFilter from the response body and set it in the webhook provider struct
+	defer resp.Body.Close()
+
+	df := endpoint.DomainFilter{}
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %v", err)
+	}
+	if err := df.UnmarshalJSON(b); err != nil {
```

Comment:
Should call `json.NewDecoder(resp.Body).Decode(&df)` to remove knowledge that JSON deserialization is done through an `UnmarshalJSON()` receiver.